### PR TITLE
Time - Kate P.

### PR DESCRIPTION
**Pull request by Catherina87**
# Assignment Submission: Hotel

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | It was challenging and difficult to incorporate hotel blocks. I had to revisit and change some of my previously created methods. 
What was a design decision you made that changed over time over the project? | I kept all my classes I initially had in the beginning. I had to get rid of some of my methods which were just duplicating each other. For example, I deleted a method select_available_room and replaced it with available_rooms_for.
What was a concept you gained clarity on, or a learning that you'd like to share? | I gained clarity on how objects can interact to each other and how to use composition. I also like that this project gave me opportunity to practice more. I also practiced method chaining that I used in the available_rooms_for method. 
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | "Room can be reserved successfully given valid arguments". It is a nominal test case because,  we are not testing the limits in this case.
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | For example, when testing an overlap? method I tested that there was no overlap when the start_date of one date range equals to the end_date of another date range. Testing the boundaries makes this test case to be an edge case one.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I think I did well, but it it was difficult to write the tests prior to writing the actual code.

### CODE REVIEW

**Comment by tildeee** [comment]
# Hotel






## Section 1: Major Learning Goals



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
Practices SRP by having at least two separate classes with distinct responsibilities, and test files for these two classes | ✔️
Overall, demonstrates understanding instance variables vs. local variables. (There aren't unnecessarily too many instance variables, when it should be a local variable) | ✔️
For each test file, tests demonstrate an understanding of instantiating objects properly, and using Arrange-Act-Assert | ✔️
Practices pseudocode and TDD, and reflected on it by filling out the reflection questions | ✔️
Practices git with at least 15 small commits and meaningful commit messages | ✔️

## Section 2: Code Review and Testing Requirements

| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
There is a class that represents a reservation, and a second class that holds/manages a collection of reservations through composition (instance variable) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date is complex logic that is separated into method(s) (and potentially class(es)) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date has unit tests | ✔️
All of the given tests run and pass | ✔️
A test coverage tool is installed and used, and shows 95% test coverage | ✔️

## Section 3: Feature Requirements

| Feature Requirement: There is a method that... | yes/no |
| --- | --- |
gives back a list of rooms, and it's tested | ✔️, yes! In `FrontDesk` it relies on the instance variable `@rooms` and the `attr_reader`, and tests with `must_responds_to` and the contents of the array
creates a specific reservation for a room for a given date range, and it has nominal test cases | ✔️
creates a specific reservation for a room for a given date range, and it tests an edge case, such as no available room, or invalid date range | Hm, I don't see an edge case test for `make_reservation` like this?
gives back a list of reservations on a given date. Its tests include a test that makes several reservations for a given date | ✔️
calculates the total price for a reservation | ✔️
gives back a list of available rooms for a given date range, and it has nominal test cases | ✔️
gives back a list of available rooms for a given date range, and it has edge test cases, such as no available room, or invalid date range | ✔️
creates a block of rooms | ✔️
reserves a room from a block | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ total in all sections | ✔️
| Yellow (Approaches Standards) | 9-13 total in all sections |
| Red (Not at Standard) | 0-8 total in all sections, or assignment is breaking/doesn’t run with less than 5 minutes of debugging |

### Additional Feedback



Kate!! You did a great job with this Hotel project! You did an amazing job of really pulling in together design, tests, and implementation, from basically nothing. Great work!


I'm really impressed with three things in particular: your comprehensive tests, your ability to find clever ways of approaching logic with Enumerable methods, and your sense of organization! It's very clear that you have a great handle on arrays and objects, and you know how to create helper methods, use different syntaxes, raise errors, etc, all to your advantage. Really great job!


I only have some minor comments for tiny improvements, but let me know if you have any questions!

## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | 
| Concise | 
| Logical/Organized | ✅

File: lib/date_range.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      if other.start_date >= @end_date
+        return false
+      elsif other.end_date <= @start_date
+        return false
+      else
+        return true
+      end
+    end
+   
+    def equals?(other)
```
Comment (tildeee) [inline lib/date_range.rb:38]: Well done: I really love that you made a method called `equals?`, because it reads really well when you call it in `FrontDesk` :)

File: lib/date_range.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    def initialize(start_date, end_date)
+      if !(start_date.is_a? Date) || !(end_date.is_a? Date)
+        raise ArgumentError.new("Invalid data type for either start or end dates.
+          Both start and end dates must be of type Date.")
+      end
+
+      if end_date < start_date
+        raise ArgumentError.new("Invalid date range: start: #{start_date}, end: #{end_date}. 
+          End date cannot be less than the start date.")
+      end
```
Comment (tildeee) [inline lib/date_range.rb:14]: Nice work here, and overall in your project, about checking validations!

File: lib/front_desk.rb [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        .filter { |reservation| reservation.date_range.overlap?(date_range) }
+        .map { |reservation| reservation.room }
+
+      rooms_in_blocks = @blocks
+        .filter { |id, block| block.date_range.overlap?(date_range) }
+        .flat_map { |id, block| block.room_collection }
+
+      available_rooms = @rooms
+        .filter { |room| !reserved_rooms.include?(room) }
+        .filter { |room| !rooms_in_blocks.include?(room) }
```
Comment (tildeee) [inline lib/front_desk.rb:24]: i LOVE the name of this method, I think it reads so well. Also, great work chaining methods!

File: lib/front_desk.rb [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    end
+
+    def reservations_by_start_date(date)
+      if !date.is_a? Date
+        raise ArgumentError.new("Invalid argument #{date}. Should be of type Date.")
+      end
+
+      @reservations.select { |reservation|
+        reservation.date_range.start_date == date
+      }
```
Comment (tildeee) [inline lib/front_desk.rb:60]: You're very much a Rubyist because here because you're relying on implicit `return`s :) But at Ada, we think you'll be set up for better success if you write more explicit `return`, so consider in the future:

```ruby
return @reservations.select { |reservation|
        reservation.date_range.start_date == date
      }
```

File: lib/front_desk.rb [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+      end
+
+      if @reservations.find { |reservation| reservation.room == room }
+        raise ArgumentError.new("Room #{room} already reserved")
+      end
+
+      @reservations << Hotel::Reservation.new(block.date_range, room)
+    end
+
+    private
```
Comment (tildeee) [inline lib/front_desk.rb:103]: YAY PRIVATE!! Great work into putting thought into what methods can be private. This is awesome.

File: lib/front_desk.rb [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+    def to_date_range(start_date, end_date)
+      return Hotel::DateRange.new(start_date, end_date)
+    end
+
+    # Took this from https://stackoverflow.com/questions/1117584/generating-guids-in-ruby
+    def generate_block_id
+      return SecureRandom.uuid
+    end
```
Comment (tildeee) [inline lib/front_desk.rb:112]: This is EXTREMELY cool. Well done!! This is awesome!!

File: test/front_desk_test.rb [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+      expect(made_reservation.last).must_be_kind_of Hotel::Reservation
+      expect(made_reservation.last.date_range.start_date).must_equal Date.new(2020, 5, 1)
+      expect(made_reservation.last.date_range.end_date).must_equal Date.new(2020, 5, 4)
+
+    end
+
+    it "adds a newly created reservation to the reservations list" do
+      front_desk = Hotel::FrontDesk.new
+      reservations_before = front_desk.reservations.length
+      puts "before length = #{reservations_before}"
```
Comment (tildeee) [inline test/front_desk_test.rb:62]: Please delete your `puts` statements before submitting your project :) It helps other people when they run your tests and code!

File: test/front_desk_test.rb [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+
+      given_date_range = Hotel::DateRange.new(Date.new(2020, 4, 10), Date.new(2020, 4, 15))
+
+      unreserved_rooms = front_desk.available_rooms_for(given_date_range)
+      expect(unreserved_rooms.length).must_equal 0
+    end
+  end
+
+  describe "#make_block" do
+    default_discount = 50
```
Comment (tildeee) [inline test/front_desk_test.rb:230]: It may make sense to refactor this into a constant variable, `DEFAULT_DISCOUNT`
